S7.Net/DataRequestItem.cs

```diff
@@ -0,0 +1,42 @@
+﻿
+using S7.Net.Types;
+
+namespace S7.Net
+{
+    internal class DataRequestItem
+    {
+        public DataRequestItem(DataType dataType, int db, int startByteAddress, int byteLength)
+        {
+            DataType = dataType;
+            DB = db;
+            StartByteAddress = startByteAddress;
+            ByteLength = byteLength;
+        }
+
+        public DataRequestItem(DataItem dataItem)
+            : this(dataItem.DataType, dataItem.DB, dataItem.StartByteAdr, Plc.VarTypeToByteLength(dataItem.VarType, dataItem.Count))
+        {
+
+        }
+
+        /// <summary>
+        /// Memory area to read 
+        /// </summary>
+        public DataType DataType { get; }
+
+        /// <summary>
+        /// Address of memory area to read (example: for DB1 this value is 1, for T45 this value is 45)
+        /// </summary>
+        public int DB { get; }
+
+        /// <summary>
+        /// Address of the first byte to read
+        /// </summary>
+        public int StartByteAddress { get; }
+
+        /// <summary>
+        /// Length of data to read
+        /// </summary>
+        public int ByteLength { get; }
+    }
```

Comment:
I'm really not a fan of this. Even though it's not publicly exposed, it adds yet another slightly different version of something that's already there. The only difference being the `ByteLength` vs the `Count`, I'd say that if all use of `DataItem` for read requests is consolidated that we don't really need this class to (Dr. Evil quotes) 'fix' a shortcoming on `DataItem`.

I doubt if I would even mind having `ByteLength` exposed on `DataItem` (one could argue the importance of knowing this `ByteLength` because it's not so much a technical _detail_ as it would be for whatever item in whatever language where addressing is not important at all). Another option is to create an extension method for the ByteLength, which could either be internal or public.

Reply:
I really did not like reusing data item for this. It does have a lot of field not relevant for building the request header, and interprets some of them quite differently as well.

I do understand that this adds yet another class, but I think that is perfectly ok because it does represent exactly the data required to create a S7 read request.

I think we should mainly discuss how we can better organize things between the different layers and classes so that something like this clearly is a internal data structure for the S7 protocol.

Oh and I definitely did not want to mess with DataItem itself.

Reply:
With Net5 this could be a record class :P
I think I will remove the constructor taking a dataItem to more clearly separate the two.

At least for me the DataItem right now is really just too much more than a mere S7 protocol data structure. It does serialization, both read and write, is mutable. Really not desirable to have for the internal stuff.

Reply:
You're absolutely right. I'm not a user (nor day to day developer) of S7NetPlus, so I didn't know about everything in `DataItem` either. I must say that records could bring welcome improvements to this codebase, also I don't think they'll be limited to Net5 either, AFAIK they're a C# 9 language feature and I don't see any reason why it wouldn't work on older targets.

All in all I'm not a fan of the name `DataRequestItem`. I agree with the structure and your vision, but maybe we can think of a better name for this? Basically this is just the address of a variable (or IO I guess, so perhaps variable is a bad term), not sure if `DataItemAddress` would rock my socks though.

Reply:
If there was some documentation of the S7 protocol we could probably name it just like that...

You are right that Request should not necessarily be in there, since that is more the function of the message sent to the PLC, and does not really define the data here.

DataItemAddress does sound quite good and in lack of a better alternative from a S7 protocol description, I will change it to that.
-------------
Assuming records do not require anything special from the runtime, they could definitely be supported for older versions. C#8 is also not enabled by default for .NET Framework, but it is still very nice to use nullable annotations.

Reply:
Are you still intending to rename this?